Speed up chat switching (and other stuff) by reducing component re-renders

marcaaron opened the pull request:
### Details
We are re-rendering a ton of stuff that doesn't really need to be re-rendered. The lists in particular are quite expensive to re-render. Also noticed that if we want to optimize chat switching over scrolling we can do a slight tweak of the FlatList props and make things much faster overall.

### Fixed Issues (Most closely related to)
https://github.com/Expensify/Expensify/issues/145841

### Tests

I would recommend testing this PR against production using the web proxy to essentially "stress test" that it is working.
I used these `.env` variables to set that up.

```
EXPENSIFY_URL_COM=https://www.expensify.com/
EXPENSIFY_URL_CASH=https://expensify.cash/
EXPENSIFY_PARTNER_NAME=chat-expensify-com
EXPENSIFY_PARTNER_PASSWORD=e21965746fd75f82bb66
PUSHER_APP_KEY=268df511a204fbb60884
NGROK_URL=https://expensify-marc.ngrok.io/
USE_NGROK=false
USE_WEB_PROXY=true
```
1. Test switching from one chat to the next
2. Verify it is faster than it was
3. Test out out switching from the Chats to Search view by tapping the magnifying glass
4. Test the FAB and the modals for New Group and New Chat and verify they are still working ok.

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [ ] Android

### Screenshots

#### Chat Switching Timing

**Before:**
<img width="441" alt="Screen Shot 2021-02-09 at 9 28 35 AM" src="https://user-images.githubusercontent.com/32969087/107417052-48ee4180-6ab9-11eb-908d-a93523aa3dcd.png">

**After:**
<img width="440" alt="Screen Shot 2021-02-09 at 9 27 37 AM" src="https://user-images.githubusercontent.com/32969087/107417098-55729a00-6ab9-11eb-8f39-606453925e36.png">

marcaaron committed:
Add various checks to optimize problem areas

marcaaron committed:
more performance improvements

marcaaron committed:
Improve chat switching time and general app performance

marcaaron committed:
add diffObject method for testing

marcaaron committed:
remove unnecessary withRouter use React.lazy on the NewGroup and NewChat views

marcaaron committed:
add activity indicator

marcaaron committed:
refactor OptionsList into class component

marcaaron committed:
use more React.lazy

marcaaron committed:
add bundle analyzer

marcaaron committed:
speed up ChatSwitcher transition

botify commented:
Hey, I noticed you changed some webpack configuration files. This can break production builds. Did you remember to run a production build locally to verify they still work?

marcaaron committed:
simplify things a bit

Julesssss commented on config/webpack/webpack.common.js:
```diff
@@ -40,6 +41,7 @@ module.exports = {
             ],
         }),
 
+        new IgnorePlugin(/^\.\/locale$/, /moment$/),
```
No harm I suppose, but we'll need to remember to remove this when we replace Moment.

marcaaron replied:
Oh yeah, so I'm not sure we're removing `moment` entirely just removing it so it doesn't produce duplicates in the main JS bundle when importing our other libraries.

marcaaron replied:
The `locales` thing just seemed unused and decided to follow this guidance 

https://github.com/jmblog/how-to-optimize-momentjs-with-webpack

marcaaron replied:
Gonna remove some of this stuff for now and focus on the bundle improvements in another issue.

Julesssss commented on src/App.js:
```diff
@@ -1,4 +1,13 @@
 
-export default () => <Expensify />;
+const App = () => (
+    <SafeAreaProvider>
+        <Expensify />
+    </SafeAreaProvider>
```
Is there a performance gain to using Safe=area-context? Took a look through the docs but didn't spot anything

marcaaron replied:
I'm not sure. We were already using this and I just moved it up a few levels since it was re-rendering when the `HomePage` re-rendered. It might be kind of cheap to do in any case, but seemed unnecessary.

Basically, I am trying to minimize the amount of stuff that re-renders when there's an update and the safe area provider is one of those things I imagine only does it's work once and never makes sense to re-render. Might also be missing some context on why we are using this vs whatever solution react-native offers.

Julesssss replied:
Ah gotcha, makes sense 👍

Julesssss commented on src/pages/NewGroupPage.js:
```diff
@@ -43,6 +43,10 @@ const propTypes = {
     ...windowDimensionsPropTypes,
 };
 
+const defaultProps = {
+    reports: {},
+};
+
```
Whats the efficiency here?

marcaaron replied:
None :) 

Just fixing the `NewGroupPage` loading with `null` reports in some cases.

marcaaron replied:
Probably will undo a bit of this. Realizing that maybe these modals don't really need to be lazy loaded.

marcaaron replied:
Gonna fix this in another issue. Some kind of bug but unrelated to these changes.

Julesssss commented on src/pages/home/HomePage.js:
```diff
@@ -44,25 +44,22 @@ import RightDockedModal from '../../components/RightDockedModal';
-import NewChatPage from '../NewChatPage';
+import FullScreenActivityIndicator from '../../components/FullScreenActivityIndicator';
+
+const NewGroupPage = lazy(() => import('../NewGroupPage'));
+const NewChatPage = lazy(() => import('../NewChatPage'));
```
Nice, this lazy loading is pretty simple!

Julesssss commented on src/pages/home/report/ReportActionItem.js:
```diff
@@ -22,4 +23,6 @@ const ReportActionItem = props => (
 ReportActionItem.propTypes = propTypes;
 
-export default ReportActionItem;
+export default memo(ReportActionItem, (prevProps, nextProps) => (
+    prevProps.displayAsGroup === nextProps.displayAsGroup && _.isEqual(prevProps.action, nextProps.action)
+));
```
Should we be worried about memory storage? Is it worth setting a limit?

marcaaron replied:
Hmm, not too sure what you mean by `maxSize`. `React.memo()` afaik does not store anything.

Julesssss replied:
Nevermind, I confused React.memo with a memoize library when researching the topic

Julesssss commented on src/pages/home/sidebar/OptionRow.js:
```diff
@@ -158,4 +158,38 @@ OptionRow.defaultProps = defaultProps;
+        return false;
+    }
+
+    if (prevProps.option.isUnread !== nextProps.option.isUnread) {
+        return false;
+    }
```
Duplicate to above condition

Julesssss reviewed (approved):
Lots of great improvements here so far. I would love to know more about how to apply changes like these to the app going forward. 

Perhaps you could share a few hints tips in Slack/Readme regarding ideal candidates for  PureComponents, memoization, and where to avoid unnecessary component updates.

Julesssss reviewed (changes requested):
Ignoring the fact this is still WIP, I'm seeing an exception in DateUtils where moment is undefined (possibly due to the `IgnorePlugin` change?)

marcaaron commented:
Nice one thanks @Julesssss I've actually got a markdown doc I'm working on that explains some of my approach here (mostly just adapted from the React docs). Will try to figure out the best way to share that.

marcaaron committed:
remove textinputfocusable changes

marcaaron committed:
undo bundle changes

marcaaron committed:
remove BundleAnalyzer for now

marcaaron committed:
remove defaultProps

marcaaron committed:
remove dupe condition and more lazy stuff

marcaaron committed:
fix conflicts

marcaaron committed:
add readme on perf

marcaaron committed:
add a bit about anonymous function

marcaaron commented:
@Julesssss I added a readme called `PERFORMANCE.md` with some of the methods I've applied here to speed things up. I've also cleaned things up so that we are only addressing re-renders here. Will look at bundle size in another issue and add more specific context there.

marcaaron committed:
fix bad speaking

marcaaron committed:
add tip to use Performance.diffObject()`

marcaaron committed:
Fix empty search

Julesssss commented:
Performance notes are a great addition
